gate-plugins/src/main/kotlin/com/netflix/spinnaker/gate/plugins/web/info/PluginInfoController.kt

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2020 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.spinnaker.gate.plugins.web.info
+
+import com.netflix.spinnaker.gate.plugins.web.PluginService
+import com.netflix.spinnaker.gate.plugins.web.SpinnakerExtensionsConfigProperties
+import com.netflix.spinnaker.gate.services.internal.Front50Service
+import io.swagger.annotations.ApiOperation
+import org.springframework.http.HttpStatus
+import org.springframework.http.MediaType
+import org.springframework.security.access.prepost.PreAuthorize
+import org.springframework.web.bind.annotation.PathVariable
+import org.springframework.web.bind.annotation.RequestBody
+import org.springframework.web.bind.annotation.RequestMapping
+import org.springframework.web.bind.annotation.RequestMethod
+import org.springframework.web.bind.annotation.RequestParam
+import org.springframework.web.bind.annotation.ResponseStatus
+import org.springframework.web.bind.annotation.RestController
+
+/**
+ * EndPoints supporting CRUD operations for PluginInfo objects.
+ */
+@RestController
+@RequestMapping("/plugin/info")
+open class PluginInfoController(
+  open val pluginService: PluginService,
+  open val front50Service: Front50Service,
+  open val spinnakerExtensionsConfigProperties: SpinnakerExtensionsConfigProperties
+) {
+
+  @ApiOperation(value = "Persist plugin metadata information")
+  @RequestMapping(
+    method = [RequestMethod.POST, RequestMethod.PUT],
+    consumes = [MediaType.APPLICATION_JSON_VALUE])
+  @ResponseStatus(value = HttpStatus.ACCEPTED)
+  fun persistPluginInfo(@RequestBody pluginInfo: Map<String, Any>): Map<String, Any> {
+    return pluginService.upsertPluginInfo(pluginInfo)
+  }
+
+  @ApiOperation(value = "Delete plugin info with the provided Id")
+  @PreAuthorize("hasPermission(#this.this.appName, 'APPLICATION', 'WRITE')")
```

Comment:
This here is the reason I had to make this class and the fields `open`.  Without those fields being open Spring throws the following when this `@PreAuthorize` annotation is present:

```
Caused by: org.springframework.aop.framework.AopConfigException: Could not generate CGLIB subclass of class com.netflix.spinnaker.gate.plugins.web.info.PluginInfoController: Common causes of this problem include using a final class or a non-visible class; nested exception is java.lang.IllegalArgumentException: Cannot subclass final class com.netflix.spinnaker.gate.plugins.web.info.PluginInfoController
```

Reply:
There is a gradle plugin for kotoin named  “allopen” that will make all classes open for spring. Keel and orca use it - would recommend using that.

Reply:
Ah, nice.  I'll check that out.

Reply:
Went with `kotlin-spring` which provides some out-of-the-box open class configuration for classes annotated with common Spring annotations.